DAOS-12449 engine: Auto-detect MD-on-SSD and LMM db path in engine

tanabarr opened the pull request:
DAOS_MD_ON_SSD and DAOS_LMM_DB_PATH environment variables currently
need to be set in the server config file in order to run MD-on-SSD.
Instead derive the necessary information from control-plane provided
daos_nvme.conf path and contents. DAOS_MD_ON_SSD_MODE should be used
to specify alternative allocator in UMEM.

- LMM db path derived from daos_nvme.conf path (engine specific
  writable dir)
- env lookups for md-on-ssd enablement replaced with 
  bio_nvme_configured(META) 

Required-githooks: true

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/1/execution/node/304/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/1/execution/node/347/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/1/execution/node/302/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/1/execution/node/357/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/1/execution/node/477/log

NiuYawei commented on src/bio/bio_xstream.c:
```diff
@@ -1116,14 +1112,15 @@ bio_nvme_configured(enum smd_dev_type type)
-	 * bdev name, that'll fail CI tests before the md on SSD feature is
-	 * fully implemented.
+	 * As wal and meta roles are not supported on the same device, single NVMe tier is not a
+	 * supported configuration for MD on SSD.
```
WAL and meta on same device is supported, and why we want to introduce the restriction (single NVMe tier isn't supported for md-on-ssd)?

tanabarr replied:
I understood from @wangshilong that WAL+meta is not currently supported https://daosio.atlassian.net/browse/DAOS-12449?focusedCommentId=109897 but if we don't want to introduce this restriction then I need to apply the role assignments explicitly in CI to provide a path to run tmpfs+single bdev tier in non-MD-on-SSD mode. This can be done.

wangshilong replied:
Sorry for misunderstaing, i thought you mean WAL+Meta query is unsupported.

NiuYawei commented on src/engine/srv.c:
```diff
@@ -1248,8 +1248,8 @@ static int
-	/* walkaround*/
-	char	*lmm_db_path = getenv("DAOS_LMM_DB_PATH");
+	/* NVMe config should always reside in an engine specific writable location */
+	char	*lmm_db_path = dirname(dss_nvme_conf);
```
This requires control plane pass in the path of dss_nvme_conf located in external storage, right?

tanabarr replied:
Yes and this is already done

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

NiuYawei commented on src/bio/bio_xstream.c:
```diff
@@ -1116,18 +1113,33 @@ bio_nvme_configured(enum smd_dev_type type)
 }
 
+int
+bio_get_lmm_db_path(char *lmm_db_path)
```
I think we could avoid exporting such an API from BIO. "lmm_db_path" looks to me is specified by caller.

tanabarr replied:
done

NiuYawei commented on src/engine/srv.c:
```diff
@@ -1261,18 +1260,20 @@ dss_sys_db_init()
-	if (lmm_db_path == NULL) {
-		D_ERROR("DAOS_LMM_DB_PATH need be configured\n");
-		return -DER_INVAL;
+	rc = bio_get_lmm_db_path(lmm_db_path);
```
Instead of calling a BIO exported API, I think we could just parse the 'dss_nvme_conf'.

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -457,7 +457,11 @@ vos_mod_init(void)
+	/**
+	 * Pass MD mode as int to enable custom umem allocator to be requested e.g. PMEM (0),
+	 * BMEM (1) and ADMEM (2). For MD on SSD specify (1) for now.
+	 */
```
Then we won't be able to enable ADMEM. I think we should retain the DAOS_MD_ON_SSD env var to specify ADMEM or BMEM (the env var name could be changed).

tanabarr replied:
done, name changed to "DAOS_MD_ON_SSD_MODE"

wangshilong commented on src/vos/vos_common.c:
```diff
@@ -776,57 +780,70 @@ vos_self_init(const char *db_path, bool use_sys_db, int tgt_id)
-		D_GOTO(failed, rc);
+		goto failed;
 
 	rc = vos_self_nvme_init(db_path, tgt_id);
```
Since you already touched this code, it is good idea to move nvme init before sys db init, otherwise, we don't know if MD-ON-SSD configured or not.

tanabarr replied:
yes good point, that needs to be called before bio_nvme_configured(META) because bio_spdk_env_init needs to be called to populate roles, thanks

tanabarr replied:
done

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -457,7 +457,11 @@ vos_mod_init(void)
+	 * Pass MD mode as int to enable custom umem allocator to be requested e.g. PMEM (0),
+	 * BMEM (1) and ADMEM (2). For MD on SSD specify (1) for now.
+	 */
+	rc = vos_pool_settings_init((int)bio_nvme_configured(SMD_DEV_TYPE_META));
```
The vos_mod_init() is called before bio_nvme_init(), bio_nvme_configured() doesn't work at this moment. (it relies on nvme_conf parsing result in bio_nvme_init()).

So we need to move this vos_pool_settings_init() call after bio_nvme_init() for both engine mode and standalone mode.

tanabarr replied:
have fixed the order in stand-alone mode (vos_self_init()) looking at how to fix the engine mode case

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -776,57 +780,70 @@ vos_self_init(const char *db_path, bool use_sys_db, int tgt_id)
-		D_GOTO(failed, rc);
+		goto failed;
+
+	if (bio_nvme_configured(SMD_DEV_TYPE_META)) {
```
The same issue here, bio_nvme_configured() doesn't work at this moment since the bio_nvme_init() isn't called yet.

Seems we need to adjust the call sequence in vos_self_init() like: bio_nvme_init() -> vos_pool_settings_init() -> sysdb init -> bio_xsctxt_alloc().

tanabarr replied:
done

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -776,57 +780,70 @@ vos_self_init(const char *db_path, bool use_sys_db, int tgt_id)
+	if (bio_nvme_configured(SMD_DEV_TYPE_META)) {
+		bool	needs_free = false;
+
+		rc = bio_get_lmm_db_path(lmm_db_path);
```
Use the passed in 'db_path'?

tanabarr replied:
and that is because in vos_self_init() there won't be an external Meta data storage path? i.e. LMM DB can live in dss_storage_path? is that why?

NiuYawei replied:
Right, to minimize code changes, I think we could store the lmm DB in the storage path for VOS unit tests and vos_perf.

tanabarr replied:
done

NiuYawei commented on src/common/mem.c:
```diff
@@ -43,18 +43,18 @@ enum {
  */
 int
-umempobj_settings_init(void)
+umempobj_settings_init(int md_mode)
```
The 'md_mode' could be changed as a 'bool' to specify if md-on-ssd is configured, and this function internally read an env var to see which allocator should be used (BMEM or ADMEM, BMEM is default) when md-on-ssd is configured.

tanabarr replied:
done, name changed to "DAOS_MD_ON_SSD_MODE"

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/2/execution/node/1212/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/3/execution/node/357/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/3/execution/node/372/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/3/execution/node/308/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/3/execution/node/319/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/3/execution/node/333/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11319/5/display/redirect

NiuYawei commented on src/common/mem.c:
```diff
@@ -38,23 +38,34 @@ enum {
+
+	if (md_on_ssd && md_mode == DAOS_MD_PMEM) {
+		D_ERROR("MD on SSD mode detected but DAOS_MD_ON_SSD set to MD_PMEM (0)");
+	}
```
Now we get "md-on-ssd enabled or not" from config file, and the "DAOS_MD_ON_SSD" env var is only used to specify "ADMEM" or "BMEM" when "md-on-ssd enabled", so this won't never happen.

tanabarr replied:
I haven't changed the behaviour of the environment variable, so the idea is that it can still be used to override the decision from the config file (for cases like unit tests). This check prints warning if we override with MD_PMEM when we have detected MD-on-SSD in the config.

tanabarr replied:
in unit tests for example the md_on_ssd will always be false, so we need a way to synthesise it

tanabarr replied:
suggested change made. Done

NiuYawei commented on src/common/mem.c:
```diff
@@ -38,23 +38,34 @@ enum {
-	d_getenv_int("DAOS_MD_ON_SSD", &val);
-	switch (val) {
+	/* Use input argument (mode detected) if not set explicitly by env var */
+	md_mode = (md_mode == DAOS_MD_INVAL) ? (int)md_on_ssd : md_mode;
```
This looks not correct to me. If md-on-ssd is configured but the env var isn't set, we should use default "BMEM", right? I think it could be like:
1. If md-on-ssd isn't configured, used "PMEM";
2. If md-on-ssd is configured, use the the env var specified backend, if env var isn't specified, use default "BMEM";
3. The env var accepts only two backends, "BMEM" or "ADMEM";

tanabarr replied:
That is what will happen? If env not set then md_mode is set to md_on_ssd (if enabled that will be BMEM, otherwise PMEM).  
If env is set then whichever mode specified will be used.
I will make your changes anyway.

NiuYawei replied:
Why do we need to override config by setting env var to PMEM? What the use case? It looks dangerous to me.

To exercise md-on-ssd in vos_tests, you need to prepare NVMe devices, when you look into the run_test.sh, you'll see that we used AIO devices for vos_tests to exercise md-on-ssd.

tanabarr replied:
Thanks for the explanations, suggested changes made. Done

NiuYawei commented on src/engine/srv.c:
```diff
@@ -1261,18 +1262,29 @@ dss_sys_db_init()
+		return -DER_NOMEM;
+	D_STRNDUP(lmm_db_path, dirname(nvme_conf_path), PATH_MAX);
+	if (lmm_db_path == NULL)
+		return -DER_NOMEM;
```
nvme_conf_path leaked.

tanabarr replied:
thanks

tanabarr replied:
done

wangshilong replied:
Maybe dirname() here change dss_nvme_conf, we still need dup string to avoid change original string unfortunately.

tanabarr replied:
I had the string copy before https://github.com/daos-stack/daos/pull/11319/commits/80ee66a77260dfcf23bccc2b3094a78532c15a79#diff-edb42428ecfead9743e3ea5d226e755fb73cb56ec4e90158bd4f5f6ded669137L1270 but was asked to remove it, I will reinstate it

NiuYawei replied:
nvme_conf_path leaked

NiuYawei commented on src/engine/srv.c:
```diff
@@ -1261,18 +1262,29 @@ dss_sys_db_init()
 		return -DER_INVAL;
 	}
 
+	D_STRNDUP(nvme_conf_path, dss_nvme_conf, PATH_MAX);
```
Why do we need to duplicate the dss_nvme_conf here?

tanabarr replied:
from manpage: "Both dirname() and basename() may modify the contents of path, so it may be desirable to pass a copy when calling one of these functions." , dss_nvme_conf is const and dirname takes char* (we can cast the const out but seems like the better behaviour is to copy so you don't have any chance of updating the original).

tanabarr replied:
have removed it anyway and just used a cast

tanabarr replied:
Done

NiuYawei commented on src/include/daos_srv/daos_engine.h:
```diff
@@ -36,6 +36,9 @@ extern unsigned int	 dss_tgt_nr;
 
+/** Number of storage tiers (hack) */
+extern unsigned int	 dss_storage_tiers;
+
```
This is not used anymore? could be removed.

tanabarr replied:
thanks

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -457,7 +457,7 @@ vos_mod_init(void)
 		vos_start_epoch = d_hlc_get();
 
-	rc = vos_pool_settings_init();
+	rc = vos_pool_settings_init(bio_nvme_configured(SMD_DEV_TYPE_META));
```
As I mentioned in prior comment, bio_nvme_configured() won't work at this moment. (when vos_mod_init() is called, the bio_nvme_init() isn't called yet). Please be aware that the engine mode is different from standalone mode, we don't call vos_self_init() for engine mode.

tanabarr replied:
how do you suggest I fix the situation, can't pass in the bool to vos_mod_init as it's a templated function

NiuYawei replied:
I think there are two ways to fix it:
1. As you proposed, move the bio_nvme_init() before dss_module_init_all();
2. Move vos_pool_settings_init() out from vos_mod_init() and put it after bio_nvme_init();

Option 1 looks cleaner to me.

tanabarr replied:
I'm made the change #1, Done

NiuYawei commented on src/vos/vos_common.c:
```diff
@@ -724,11 +724,6 @@ vos_self_nvme_init(const char *vos_path, uint32_t tgt_id)
-	if (rc) {
-		D_ERROR("Failed to allocate NVMe context. "DF_RC"\n", DP_RC(rc));
-		goto out;
-	}
```
This his removed from vos_self_nvme_init() but the vos_self_nvme_fini() isn't updated accordingly?

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/d1d6f5bf45348e9ab7c37903f16985c7da48bd35/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/f5c9e06b76ad32664bd7b5003143a67c62d38447/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/c030a24f506b7cf4420e686a78c090a0139eb398/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/bc68d528ba1ed00c960aed0785158433fc49b250/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11319/10/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/8176fc065de075479b51c3dbbf6a9a6dae842c26/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/11/execution/node/326/log

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/11/execution/node/372/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/11/execution/node/317/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/8176fc065de075479b51c3dbbf6a9a6dae842c26/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/12/execution/node/327/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/12/execution/node/373/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/12/execution/node/303/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/12/execution/node/300/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/12/execution/node/349/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/0f443e7c779f7b260ed1e105e47a40c9000fb5ba/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/290629df3a5f744ee2d34ff313d8a50cf972a77e/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/14/execution/node/349/log

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/14/execution/node/373/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/14/execution/node/319/log

tanabarr commented:
Unfortunately the reordering seems to have caused a problem on initialisation when MD-on-SSD is enabled:
```
02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] common INFO src/common/mem.c:73 umempobj_settings_init() UMEM will use Blob Backed Memory as the metadata backend interface
02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] mgmt DBUG src/mgmt/srv.c:424 ds_mgmt_init() successful init call
02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/init.c:776 server_init() Module vos,rdb,rsvc,security,mgmt,dtx,pool,cont,obj,rebuild successfully initialized
02/02-17:04:37.11 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/srv.c:978 dss_xstreams_init() CPU relax interval is set to 1 msecs
02/02-17:04:37.11 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/srv.c:990 dss_xstreams_init() CPU relax mode is set to [net]
02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  INFO src/bio/bio_xstream.c:1538 bio_xsctxt_alloc() Initialize NVMe context, tgt_id:1024, init_thread:(nil)
02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  ERR  src/bio/bio_xstream.c:1576 bio_xsctxt_alloc() failed to init bdevs, rc:-1003
02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  DBUG src/bio/bio_xstream.c:1483 bio_xsctxt_free() Finalizing SPDK thread, tgt_id:1024
02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  DBUG src/bio/bio_xstream.c:1490 bio_xsctxt_free() SPDK thread finalized, tgt_id:1024
02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/srv.c:468 dss_srv_handler() failed to init spdk context for xstream(0) rc:-1003
02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/sched.c:1981 sched_watchdog_post() WATCHDOG: Thread 0x564621ef8160 took 269 ms. symbol:/home/tanabarr/projects/daos/install-rocky/bin/daos_engine(+0x4a160) [0x564621ef8160]
02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/init.c:784 server_init() DAOS cannot be initialized using the configured path (/mnt/daos0).   Please ensure it is on a PMDK compatible file system and writeable by the current user
02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] mgmt DBUG src/mgmt/srv.c:433 ds_mgmt_fini() successful fini call
```

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/14/execution/node/325/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/14/execution/node/324/log

wangshilong commented:
> Unfortunately the reordering seems to have caused a problem on initialisation when MD-on-SSD is enabled:
> 
> ```
> 02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] common INFO src/common/mem.c:73 umempobj_settings_init() UMEM will use Blob Backed Memory as the metadata backend interface
> 02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] mgmt DBUG src/mgmt/srv.c:424 ds_mgmt_init() successful init call
> 02/02-17:04:37.08 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/init.c:776 server_init() Module vos,rdb,rsvc,security,mgmt,dtx,pool,cont,obj,rebuild successfully initialized
> 02/02-17:04:37.11 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/srv.c:978 dss_xstreams_init() CPU relax interval is set to 1 msecs
> 02/02-17:04:37.11 wolf-226 DAOS[3078135/-1/0] server INFO src/engine/srv.c:990 dss_xstreams_init() CPU relax mode is set to [net]
> 02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  INFO src/bio/bio_xstream.c:1538 bio_xsctxt_alloc() Initialize NVMe context, tgt_id:1024, init_thread:(nil)
> 02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  ERR  src/bio/bio_xstream.c:1576 bio_xsctxt_alloc() failed to init bdevs, rc:-1003
> 02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  DBUG src/bio/bio_xstream.c:1483 bio_xsctxt_free() Finalizing SPDK thread, tgt_id:1024
> 02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] bio  DBUG src/bio/bio_xstream.c:1490 bio_xsctxt_free() SPDK thread finalized, tgt_id:1024
> 02/02-17:04:37.37 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/srv.c:468 dss_srv_handler() failed to init spdk context for xstream(0) rc:-1003
> 02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/sched.c:1981 sched_watchdog_post() WATCHDOG: Thread 0x564621ef8160 took 269 ms. symbol:/home/tanabarr/projects/daos/install-rocky/bin/daos_engine(+0x4a160) [0x564621ef8160]
> 02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] server ERR  src/engine/init.c:784 server_init() DAOS cannot be initialized using the configured path (/mnt/daos0).   Please ensure it is on a PMDK compatible file system and writeable by the current user
> 02/02-17:04:37.38 wolf-226 DAOS[3078135/-1/0] mgmt DBUG src/mgmt/srv.c:433 ds_mgmt_fini() successful fini call
> ```

I did some debugging, it looks reason is you need pass correct dss_nvme_conf here, my output here is: 
02/03-09:48:42.48 daos1 DAOS[1032225/-1/0] bio  ERR  src/bio/bio_xstream.c:1576 bio_xsctxt_alloc() failed to init bdevs, rc:-1003, conf: /mnt/daos0

So please it should be /mnt/daos0/daos_nvme.conf?

NiuYawei commented on src/engine/srv.c:
```diff
@@ -1261,18 +1257,24 @@ dss_sys_db_init()
 		return -DER_INVAL;
 	}
 
+	D_STRNDUP(lmm_db_path, dirname((char *)dss_nvme_conf), PATH_MAX);
```
I guess 'dss_nvme_conf' is truncated by this dirname() call, that caused failures.

tanabarr replied:
I had the string copy before https://github.com/daos-stack/daos/pull/11319/commits/80ee66a77260dfcf23bccc2b3094a78532c15a79#diff-edb42428ecfead9743e3ea5d226e755fb73cb56ec4e90158bd4f5f6ded669137L1270 but was asked to remove it, I will reinstate it

tanabarr replied:
https://github.com/daos-stack/daos/pull/11319#discussion_r1092018870

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/4d3c6fe26e43c47c50b5c50a9a68eecb7f20a3ad/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/15/execution/node/370/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/15/execution/node/372/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/15/execution/node/313/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/15/execution/node/319/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/15/execution/node/324/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/459da39915a43c111eb354ee409479f1801e736f/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/16/execution/node/311/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/16/execution/node/324/log

tanabarr commented:
Pool create now working with MD-on-SSD BMEM mode after fixing conf path truncation issue thanks @NiuYawei and @wangshilong .
Removed env use for vos_tests in run_test (DAOS_MD_ON_SSD_MODE only needed to test ADMEM instead of BMEM now), and updated src/vos/tests/daos_nvme.conf to apply bdev role assignments which will trigger BMEM mode in test.

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/16/execution/node/305/log

daosbuild1 commented:
Test stage Unit Test on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/16/execution/node/622/log

daosbuild1 commented:
Test stage Unit Test with memcheck on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/16/execution/node/624/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/4d70646ce5b0f44568c323d4adca77e878286a6a/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/17/execution/node/321/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/17/execution/node/318/log

NiuYawei commented on utils/run_test.sh:
```diff
@@ -150,9 +150,7 @@ if [ -d "/mnt/daos" ]; then
         dd if=/dev/zero of="${AIO_DEV}" bs=1G count=20
         sed -i "s+\"name\": \"AIO_1\"+\"name\": \"AIO_7\"+g" ${NVME_CONF}
```
This could be removed since you updated the dev name in daos_nvme.conf.

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/17/execution/node/357/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/e7489a928309cff3ca3214bfa548d0174acd7ea0/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/e7489a928309cff3ca3214bfa548d0174acd7ea0/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/19/execution/node/347/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11319/19/execution/node/291/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/5c7ff9195b74a532467ef55903ba7af2060e4cbd/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/5c7ff9195b74a532467ef55903ba7af2060e4cbd/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]

jolivier23 commented on utils/run_test.sh:
```diff
@@ -148,17 +148,11 @@ if [ -d "/mnt/daos" ]; then
 
         rm -f "${AIO_DEV}"
         dd if=/dev/zero of="${AIO_DEV}" bs=1G count=20
```
Does this still test MD_ON_SSD?   It looks identical to the run on 147 except with bigger AIO file.  Previously, it was using PMDK for one and NVMD_ON_SSD for the other.

jolivier23 reviewed (dismissed):
ok to defer fixing the PMDK test to another patch

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/run_test.sh:84](https://github.com/daos-stack/daos/blob/2e0b1e9d59cc7cb23218141db42f147c509cca4a/utils/run_test.sh#L84):
(lint) eval negates the benefit of arrays. Drop eval to preserve whitespace/symbols (or eval as string). [SC2294]
